### PR TITLE
ES-7: remove type

### PR DESCRIPTION
**Pull request by thePanz**
Todo:
- [x] Remove mapping's  include_type_name (from #1654 )
- [x] Deprecate usage of Type object
- [x] copy `setMapping()` from `Type` to `Index`
- [x] copy `getDocument()` from `Type` to `Index`
- [x] copy `deleteById() from `Type` to `Index`
- [x] copy more methods from `Type` to `Index`
- [x] Cleanup `Client` methods requiring `Type` #1668
- [x] Remove `Type` class (<== needs clarification)
- [x] Cleanup `Search->addType()` #1669
- [x] Cleanup `Search->addTypes()` #1669
- [x] Cleanup `Bulk` class from `Type` #1670
- [x] Fix tests

### CODE REVIEW

**Comment by thePanz** [comment]
@ruflin the usage of ` include_type_name` was too hard to remove while keeping a backward-compatibility. This PR is a first step to removing the `Type` completely, and moving its functionalities into the `Index` class.

**Comment by renanbr** [comment]
Assuming everything is fine in `src/` I changed some tests.
Let me know if it's in the good direction.

Changes I made, but not 100% convinced:

- `Elastica\Test\BulkTest::testSend()` and `Elastica\Test\BulkTest::testRawDocumentDataRequest()`
    - orpahend document (has no index) is always tied to the type "_doc"

Wouldn't we deprecate (and auto-redirect when object is given) also these methods ?

- `Elastica\Bulk->setType()` ==> `Elastica\Bulk->setIndex()`
- `Elastica\Search->addType()` ==> `Elastica\Search->addIndex()`
- `Elastica\Search->addTypes()` ==> `Elastica\Search->addIndices()`

Additional note:

- `Elastica\Client::deleteIds()` and `Elastica\Client::updateDocument()` still require `$type` argument, is this intentional?

**Comment by thePanz** [comment]
@renanbr thank you for your findings, I updated the PR description

**Comment by renanbr** [comment]
@thePanz I've submitted some PRs:

- [ES-7] Make `Client` class `Type` free #1668
- [ES-7] Cleanup `Search->addType()` and `Search->addTypes()` #1669
- [ES-7] Make `Bulk` classes `Type` free #1670 (Bulk weren't caught)

Let me know if they are in the good direction.

The missing tasks are:

- Remore `Type` class
- Fix tests\*

\* tests for these classes I've changed are green :)

**Comment by thePanz** [comment]
@renanbr I'll have a look asap.

I think that all your changes should be merged into this MR, to have all "remove Type class" changes in one single pull-request. wdyt?

File: lib/Elastica/Index.php [hunk 1]
```diff
@@ -194,13 +255,84 @@ public function addDocuments(array $docs, array $options = [])
...
+
+        return $document;
+    }
+
+    /**
+     * Deletes a document by its unique identifier.
+     *
+     * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete.html
+     *
+     * @param int|string $id      Document id
```
Comment (alamirault) [inline lib/Elastica/Index.php]: just string

File: lib/Elastica/Index.php [hunk 2]
```diff
@@ -618,4 +746,37 @@ public function analyze(array $body, $args = [])
...
+    {
+        return new Type($this, Document::DEFAULT_TYPE);
+    }
+
+    /**
+     * Update document, using update script.
+     *
+     * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html
+     *
+     * @param Document|AbstractScript $data    Document or Scriwith update data
```
Comment (alamirault) [inline lib/Elastica/Index.php]: Script with ?

File: lib/Elastica/Search.php [hunk 3]
```diff
@@ -410,29 +338,11 @@ public function getPath()
             return '_search/scroll';
         }
```
Comment (alamirault) [inline lib/Elastica/Search.php]: Php doc must be "Combines indices to the search request path."

**Comment by renanbr** [comment]
more test fixes #1674 (targeting this branch, as usual)

**Comment by renanbr** [comment]
Tests are green :smile:
Is there still a task to be done before getting this PR merged?
Is there a plan to release a major version compatible with ES 7?

**Comment by ruflin** [comment]
@renanbr Main thing missing is me getting back from PTO and catching up on things. 7.0 release is planned yes. Hope to get all these in this and next week 🤞

**Comment by renanbr** [comment]
Thanks for the quick response! Let me know if you need some more hand :wink:

**Comment by ruflin** [comment]
@thePanz @renanbr Didn't get to fully review this yet but one more generic question: If a user has a type `foo` in an index coming from Elasticsearch 6.x. Will Elastica still offer a way to query this index?

**Comment by renanbr** [comment]
Shouldn't we notify that `Document`'s constructor casts id to `string` ? and also change `AbstractUpdateAction::setId()` and `AbstractUpdateAction::getId()` methods signatures?

**Comment by thePanz** [comment]
@ruflin if I get the documentation correctly, indexes created in ES-6, can be used in ES-7, even if the 6.x version had a (single) type.
The documentation I found is here: https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html#_typeless_apis_in_7_0

WDYT?

**Comment by ruflin** [comment]
@thePanz Yes. AFAIK it works directly in Elasticsearch, but will a user still be able to query his index with type `foo` in Elastica 7?

**Comment by thePanz** [comment]
@ruflin from the documentation I get that ES will use the only available index "type" in ES-7. That means that if the ES-7 index was migrated from ES-6, with a `foo` type, it will be used automatically by ES-7, as new code changes removed the `_doc` type.
It should be double checked tho

**Comment by ruflin** [comment]
@thePanz Good to know. That is the part I'm also not 100% sure about but I hope it works this way. If it does, we are save :-)

File: lib/Elastica/Document.php [hunk 4]
```diff
@@ -36,16 +37,14 @@ class Document extends AbstractUpdateAction
...
      * Creates a new document.
      *
-     * @param int|string   $id    OPTIONAL $id Id is create if empty
+     * @param string       $id    OPTIONAL $id Id is create if empty
      * @param array|string $data  OPTIONAL Data array
-     * @param Type|string  $type  OPTIONAL Type name
      * @param Index|string $index OPTIONAL Index name
      */
-    public function __construct($id = '', $data = [], $type = '', $index = '')
+    public function __construct(string $id = null, $data = [], $index = '')
```
Comment (ruflin) [inline lib/Elastica/Document.php]: string and `null`?

File: lib/Elastica/Index.php [hunk 5]
```diff
@@ -100,23 +89,32 @@ public function getRecovery()
...
 
         // Get first entry as if index is an Alias, the name of the mapping is the real name and not alias name
         $mapping = \array_shift($data);
 
-        if (isset($mapping['mappings'])) {
-            return $mapping['mappings'];
-        }
-
-        return [];
+        return $mapping['mappings'] ?? [];
```
Comment (ruflin) [inline lib/Elastica/Index.php]: Is this manual cleanup or some cleanup your IDE does automatically?

File: lib/Elastica/Result.php [hunk 6]
```diff
@@ -74,6 +74,8 @@ public function getId()
      * Returns the type of the result.
      *
      * @return string Result type
+     *
+     *@deprecated ES 7.x deprecated the use of types in the index
```
Comment (ruflin) [inline lib/Elastica/Result.php]:
```suggestion
     * @deprecated ES 7.x deprecated the use of types in the index
```

**Comment by ruflin** [comment]
@thePanz @philkra @renanbr For me, the potential issue for users which migrated from users from 6.x to 7 with a non standard type is kind of a blocker for this PR. Unfortunately did not get yet to test it if it is as suggested here: https://github.com/ruflin/Elastica/pull/1666#discussion_r338072603

Anyone time to manually test it? Should we hop on a call to discuss this further? Any way moving forward with this PR but have an escape path for users with this problem (like a hack 😇 )?

**Comment by philkra** [comment]
@thePanz @ruflin @renanbr , I tested the upgrade manually the following:
* spun up an es cluster `v6.8.3`
* ingested documents into an index `twitter` and a type `my_type`
* upgraded es to `v7.4.0`
I was able to pull the documents with `$client->getIndex('twitter')->getDocument($docId)` as well as to add a document `$client->getIndex('twitter')->addDocument($id, $doc)`.
In other words my fear of locking out everyone with an upgraded mapping type was not backed by evidence :)

**Comment by thePanz** [comment]
Thank you @philkra , I guess the documentation here: https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html#_typeless_apis_in_7_0 is right then :)

File: lib/Elastica/Document.php [hunk 7]
```diff
@@ -12,7 +12,8 @@
  */
 class Document extends AbstractUpdateAction
 {
-    const OP_TYPE_CREATE = Action::OP_TYPE_CREATE;
+    public const OP_TYPE_CREATE = Action::OP_TYPE_CREATE;
+    public const DEFAULT_TYPE = '_doc';
```
[the remark this answers is not in the document]
Comment (thePanz) [reply lib/Elastica/Document.php]: It should work to get your custom types: see https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html#_typeless_apis_in_7_0

**Comment by philkra** [comment]
@thePanz , I was rather cautious because all language clients still offer the definition of `type` .. but just to rephrase: `GET twitter/my_type/1` is the same as `GET twitter/_doc/1` in ES 7.0 by definition of typeless API's. This blocker is then obsolete.

**Comment by renanbr** [comment]
there are and some more php-doc fixes https://github.com/renanbr/Elastica/commit/41315e5142b15f38d05444686ca57b7ec590abde

**Comment by thePanz** [comment]
@renanbr merged!

File: lib/Elastica/Index.php (hunk 5 above)
Comment (thePanz) [reply lib/Elastica/Index.php]: My IDE highlights the code and suggests some optimizations :)

File: lib/Elastica/Document.php (hunk 4 above)
Comment (thePanz) [reply lib/Elastica/Document.php]: True, as the id will be generated if not given

**Comment by ruflin** [comment]
@philkra Ok, just to make sure we are on the same page. So even if a a user migrated with  a type `foo` from 6.x to 7, all the index actions will work even he can't set `foo` as type? If yes, that would be greatè

**Comment by philkra** [comment]
@ruflin , yes correct! The `type` is part in an endpoint URL is essentially just a placeholder, so using `_doc` with every request is sound.

**Comment by renanbr** [comment]
> @ruflin , yes correct! The `type` part in an endpoint URL is essentially just a placeholder in `v7.x`, so using `_doc` with every request is sound.

Be aware that this PR doesn't put `_doc` as type explicitly, **type are actually omitted**, what maybe makes this PR work even with named types from ES6

File: lib/Elastica/Aggregation/Children.php [hunk 8]
```diff
@@ -5,14 +5,14 @@
 /**
  * Class Children.
  *
- * @see https://www.elastic.co/guide/en/elasticsearch/reference/1.7/search-aggregations-bucket-children-aggregation.html
+ * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-children-aggregation.html
```
Comment (ruflin) [inline lib/Elastica/Aggregation/Children.php]: 1.7 is not current anymore 🤔 😆

File: lib/Elastica/Query/ParentId.php [hunk 9]
```diff
@@ -12,29 +12,29 @@ class ParentId extends AbstractQuery
...
-     * @param bool       $ignoreUnmapped
+     * @param string $type           Name of the child relationship mapped for the join field
+     * @param string $id             ID of the parent document. The query will return child documents of this parent document.
+     * @param bool   $ignoreUnmapped Indicates whether to ignore an unmapped type and not return any documents instead of an error. Defaults to false.
      */
-    public function __construct(string $type, $id, bool $ignoreUnmapped = false)
+    public function __construct(string $type, string $id, bool $ignoreUnmapped = false)
     {
-        $this->setType($type);
+        $this->setRelationshipType($type);
```
Comment (ruflin) [inline lib/Elastica/Query/ParentId.php]: Haven't check this yet, but I assume this still exists?

File: lib/Elastica/Reindex.php [hunk 10]
```diff
@@ -41,11 +40,6 @@ class Reindex extends Param
      */
     protected $_newIndex;
 
-    /**
-     * @var array
-     */
-    protected $_options;
```
Comment (ruflin) [inline lib/Elastica/Reindex.php]: Was this not used anymore anyways?

File: test/Elastica/IndexTest.php [hunk 11]
```diff
@@ -835,29 +815,6 @@ public function testAnalyzeExplain()
         $this->assertArrayHasKey('custom_analyzer', $data);
     }
 
-    /**
-     * @group unit
-     */
-    public function testThrowExceptionIfNotScalar()
```
Comment (ruflin) [inline test/Elastica/IndexTest.php]: Not sure who the removal of this is related? Not needed anymore?

File: test/Elastica/MappingTest.php [hunk 12]
```diff
@@ -322,27 +299,21 @@ public function testSetMeta()
...
-        $type = $index->getType('_doc');
         $properties = [
             'firstname' => ['type' => 'text', 'store' => true],
             'lastname' => ['type' => 'text'],
         ];
-        $mapping = new Mapping($type, $properties);
+        $mapping = new Mapping($properties);
         $all = [
-           'enabled' => true,
+            'enabled' => true,
```
Comment (ruflin) [inline test/Elastica/MappingTest.php]: Nit: not sure where this ` ` is coming from.

**Comment by philkra** [comment]
> Be aware that this PR doesn't put _doc as type explicitly, type are actually omitted, what maybe makes this PR work even with named types from ES6

@renanbr the type may be omitted as param, but it is added to the URI by the [`elasticsearch-php`](https://github.com/elastic/elasticsearch-php) client. The abstract method  of `AbstractEndpoint::getURI()` is always adding `_doc` as type if none provided, for example [Index:L42](https://github.com/elastic/elasticsearch-php/blob/master/src/Elasticsearch/Endpoints/Index.php#L42).

**Comment by philkra** [comment]
@ruflin , I'm 👍 on this PR and can only second the amazement of the gigantic effort that went in to this PR!!

File: lib/Elastica/Document.php (hunk 7 above)
Comment (philkra) [inline lib/Elastica/Document.php]: Could `DEFAULT_TYPE` may be removed as it's not used?

**Comment by ruflin** [comment]
@thePanz Will also need a rebase.

File: lib/Elastica/Query/ParentId.php (hunk 9 above)
Comment (thePanz) [reply lib/Elastica/Query/ParentId.php]: Don't think so :)

Comment (thePanz) [reply lib/Elastica/Query/ParentId.php]: this is just a rename of the private method, is "type" here is for the relationship type

File: lib/Elastica/Reindex.php (hunk 10 above)
Comment (thePanz) [reply lib/Elastica/Reindex.php]: yes, not used anymore

File: lib/Elastica/Document.php (hunk 7 above)
Comment (thePanz) [reply lib/Elastica/Document.php]: Removed!

**Comment by thePanz** [comment]
@ruflin rebased and pushed!

File: test/Elastica/IndexTest.php (hunk 11 above)
Comment (thePanz) [reply test/Elastica/IndexTest.php]: As we now use PHP typehings, passing something that is not a scalar will throw a PHP error :)

**Comment by ruflin** [comment]
Merged 🎉 🎈

**Comment by ruflin** [comment]
Just created https://github.com/ruflin/Elastica/pull/1685 to prepare the 7.0.0-beta1 release.

**Pull request merged**